processing/src/main/java/org/apache/druid/query/expression/IPv4AddressParseExprMacro.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.druid.math.expr.Expr;
+import org.apache.druid.math.expr.ExprEval;
+import org.apache.druid.math.expr.ExprMacroTable;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.net.Inet4Address;
+import java.net.InetAddress;
+import java.util.List;
+
+/**
+ * <pre>
+ * Implements an expression that parses string or long into an IPv4 address stored (as an unsigned
+ * int) in a long.
+ *
+ * Expression signatures:
+ * - long ipv4address_parse(string)
+ * - long ipv4address_parse(long)
+ *
+ * Valid argument formats are:
+ * - IPv4 address dotted-decimal notation string (e.g., "198.168.0.1")
+ * - IPv6 IPv4-mapped adress string (e.g., "::ffff:192.168.0.1")
+ * - unsigned int long (e.g., 3232235521)
+ * - unsigned int string (e.g., "3232235521")
+ *
+ * Invalid arguments return null.
+ *
+ * The overloaded signature allows applying the expression to a dimension with mixed string and long
+ * representations of IPv4 addresses.
+ * </pre>
+ *
+ * @see IPv4AddressStringifyExprMacro
+ * @see IPv4AddressMatchExprMacro
+ */
+public class IPv4AddressParseExprMacro implements ExprMacroTable.ExprMacro
+{
+  public static final String NAME = "ipv4address_parse";
+
+  @Override
+  public String name()
+  {
+    return NAME;
+  }
+
+  @Override
+  public Expr apply(final List<Expr> args)
+  {
+    if (args.size() != 1) {
+      throw new IAE("Function[%s] must have 1 argument", name());
+    }
+
+    Expr arg = args.get(0);
+
+    class IPv4AddressParseExpr extends ExprMacroTable.BaseScalarUnivariateMacroFunctionExpr
+    {
+      private IPv4AddressParseExpr(Expr arg)
+      {
+        super(arg);
+      }
+
+      @Nonnull
+      @Override
+      public ExprEval eval(final ObjectBinding bindings)
+      {
+        String stringValue = arg.eval(bindings).asString();
+        if (stringValue == null) {
+          return ExprEval.ofLong(null);
+        }
+
+        // Assume use cases in order of most frequent to least are:
+        // 1) convert string to long
+        // 2) convert long to long
+        Long value = parseAsString(stringValue);
+        if (value == null) {
+          value = parseAsLong(stringValue);
+        }
+
+        return ExprEval.ofLong(value);
+      }
+
+      @Override
+      public Expr visit(Shuttle shuttle)
+      {
+        Expr newArg = arg.visit(shuttle);
+        return shuttle.visit(new IPv4AddressParseExpr(newArg));
+      }
+    }
+
+    return new IPv4AddressParseExpr(arg);
+  }
+
+  @Nullable
+  private static Long parseAsString(String stringValue)
+  {
+    try {
+      // Do not use java.lang.InetAddress#getByName() as it may do DNS lookups
+      InetAddress address = InetAddresses.forString(stringValue);
+      if (address instanceof Inet4Address) {
+        int value = InetAddresses.coerceToInteger(address);
+        return Integer.toUnsignedLong(value);
+      }
+    }
+    catch (IllegalArgumentException ignored) {
+      // fall through (Invalid IPv4 adddress string)
```

Comment:
Naively this feels like it could be very expensive to eat an exception every row we evaluate an expression for the case where all values in the column will fall through to `parseAsLong`. Is there a less explodey way to check if the string is chill?

Reply:
One option is to compare against a regular expression to validate the format before trying to parse.

Reply:
Another option is [`InetAddresses.isInetAddress()`](https://guava.dev/releases/19.0/api/docs/com/google/common/net/InetAddresses.html#isInetAddress(java.lang.String))